engine/src/test/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStoreSpec.scala

```diff
@@ -0,0 +1,60 @@
+package cromwell.engine.workflow.lifecycle.execution.stores
+
+import cromwell.backend.BackendJobDescriptorKey
+import cromwell.core.ExecutionStatus.{ExecutionStatus, _}
+import cromwell.core.JobKey
+import cromwell.engine.workflow.lifecycle.execution.stores.ExecutionStoreSpec._
+import org.scalatest.{BeforeAndAfter, FlatSpec, Matchers}
+import wom.graph._
+
+import scala.util.Random
+
+class ExecutionStoreSpec extends FlatSpec with Matchers with BeforeAndAfter {
+
+  var store: ExecutionStore = _
+
+  before {
+    def jobKeys: Map[JobKey, ExecutionStatus] = (0.until(10000).toList map {
+      i => BackendJobDescriptorKey(noConnectionsGraphNode, Option(i), 1) -> NotStarted })
+      .toMap
+
+    store = ActiveExecutionStore(jobKeys, needsUpdate = true)
+  }
+
+  def updateStoreToEnqueueNewlyRunnableJobs(): Unit = {
+    while (store.needsUpdate) {
+      val update = store.update
+      store = update.updatedStore.updateKeys(update.runnableKeys.map(_ -> QueuedInCromwell).toMap)
+    }
+  }
+
+  it should "keep allowing updates while 10000 call keys are enqueued and then started in small batches" in {
+
+    updateStoreToEnqueueNewlyRunnableJobs()
+
+    store.store.getOrElse(QueuedInCromwell, List.empty).size should be(10000)
+    store.store.getOrElse(Running, List.empty).size should be(0)
+
+    while(store.store.getOrElse(Running, List.empty).size < 10000) {
```

Comment:
Looks like in case of some mistake in ExecutionStore's logic this may become an endless loop. I'd add some logic inside the loop to check that if number of running jobs doesn't change between any 2 iterations, then it's a bug.